### PR TITLE
Move environment configuration to its own class

### PR DESCRIPTION
**Pull request by b-butler**
## Description
Adds three classes
- `_PartitionConfig`: an environment's partition configuration
- `_Partition`: a specific partition's data
- `_NodeTypes`: what kind of nodes a partition supports
  - SHARED: 1 or fewer nodes
  - MIXED: any number of nodes/processes
  - WHOLENODE: submit in whole nodes only

## Motivation and Context
With recent changes to flow's template/submission logic (see #722), the environment classes were becoming loaded with class variables. This makes a logical change to storing the partition specific information in appropriate classes. This also removes some methods from the environments in favor of putting them in the partition classes.

## Checklist:
- [x] I am familiar with the [Contributing Guidelines](https://github.com/glotzerlab/signac-flow/blob/main/CONTRIBUTING.md).
- [x] I agree with the terms of the [Contributor Agreement](https://github.com/glotzerlab/signac-flow/blob/main/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/main/contributors.yaml).
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.
- [x] The [package documentation](https://github.com/glotzerlab/signac-flow/tree/main/doc) and [framework documentation](https://docs.signac.io/) in [signac-docs](https://github.com/glotzerlab/signac-docs) are up to date with these changes.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/main/changelog.txt) and added any related issue and pull request numbers for future reference.

### CODE REVIEW

**Comment by b-butler** [comment]
Tested appropriate errors for GPU/non-GPU and node requests on Bridges2.

**Comment by b-butler** [comment]
@cbkerr when you have a moment could you review?

**Comment by b-butler** [comment]
@cbkerr will you have time to review or should I assign someone else?

**Comment by cbkerr** [comment]
@b-butler Thanks for the ping. I can get to this during the next few days.

File: flow/environment.py [hunk 1]
```diff
@@ -90,6 +91,169 @@ def template_filter(func):
...
+    - CPUs for a partition
+    - GPUs for a partition
+    - Node type of a partition
+
+    When querying a value for a specific partition the logic first searches the
+    provided mapping if any for the partition, if it is not found then the
+    mapping is searched for "default" if it exists, if not the class default is
+    used.
+
+    1. Partition specific -> 2. Provided default -> 3. _PartitionConfig default
```
Comment (cbkerr) [inline flow/environment.py:122]: What do the numbers represent? If search order, put them in a list instead

File: flow/environments/umich.py [hunk 2]
```diff
@@ -13,10 +13,10 @@ class GreatLakesEnvironment(DefaultSlurmEnvironment):
 
     hostname_pattern = r"gl(-login)?[0-9]+\.arc-ts\.umich\.edu"
     template = "umich-greatlakes.sh"
-    _cpus_per_node = {"default": 36, "gpu": 40}
-    _gpus_per_node = {"default": 2}
-    _shared_partitions = {"standard", "gpu"}
```
Comment (cbkerr) [inline flow/environments/umich.py:18]: What happened to the standard partition?

File: flow/environment.py [hunk 3]
```diff
@@ -90,6 +91,169 @@ def template_filter(func):
...
+
+    - CPUs: ``None`` which represents any number. This will be interpretted as
+      an unlimited supply of CPUs practically.
+    - GPUs: 0.
+    - Node type: `_NodeTypes.MIXED`.
+
+    Parameters
+    ----------
+    cpus_per_node: dict[str, int], optional
+        Mapping between partitions and CPUs per node. Defaults to an empty `dict`.
```
Comment (cbkerr) [inline flow/environment.py:134]: I had to look at the environments to understand this docstring. Need to mention that keys in this dictionary are partition names.

I understand why "mapping" was used, but I couldn't figure out which direction it was pointing and how to populate the dictionary based on this info.

```suggestion
        A `dict` mapping from partition names (keys) to CPUs per node (values) on those partitions.
        Defaults to an empty `dict`.
```

Also applies to the other docstrings but we can discuss here first

Comment (b-butler) [reply flow/environment.py:134]: A mapping from a to b means a are keys and b are values.

File: flow/environments/umich.py (hunk 2 above)
Comment (b-butler) [reply flow/environments/umich.py:18]: The standard partition as well as the gpu partition don't need specifying since their behavior matches the default.

**Review by cbkerr — APPROVED** [review]
Concerning documentation:

Are there any user-facing changes to the template and environment documentation we would need to add?

What are the risks of default values changing and messing up inherited values?

**Comment by b-butler** [comment]
@cbkerr Nope, this is purely a developer facing thing currently.
And the way the model works is that the defaults are inherited as class members from `_ComputeEnvironment` but you can't partially inherit meaning you must explicitly set the full specification for every new class.

**Pull request merged**